services/node/src/server.rs

```diff
@@ -0,0 +1,195 @@
+use async_trait::async_trait;
+use common::*;
+use log::info;
+use mbus_api::*;
+use std::{collections::HashMap, convert::TryInto, marker::PhantomData};
+use structopt::StructOpt;
+use tokio::sync::Mutex;
+
+#[derive(Debug, StructOpt)]
+struct CliArgs {
+    /// The Nats Server URL to connect to
+    /// (supports the nats schema)
+    /// Default: nats://127.0.0.1:4222
+    #[structopt(long, short, default_value = "nats://127.0.0.1:4222")]
+    url: String,
+    /// Deadline for the mayastor instance keep alive registration
+    /// Default: 2
+    #[structopt(long, short, default_value = "2s")]
+    deadline: humantime::Duration,
+}
+
+/// Needed so we can implement the ServiceSubscriber trait for
+/// the message types external to the crate
+#[derive(Clone, Default)]
+struct ServiceHandler<T> {
+    data: PhantomData<T>,
+}
+
+#[derive(Clone)]
+struct Watchdog {
+    deadline: std::time::Duration,
+    pet_chan: tokio::sync::mpsc::Sender<()>,
+}
+
+impl Watchdog {
+    /// new empty watchdog with a timeout
+    pub fn new(deadline: std::time::Duration) -> Self {
+        Self {
+            deadline,
+            pet_chan: tokio::sync::mpsc::channel(1).0,
+        }
+    }
+
+    /// arm watchdog with self timeout and execute error callback if
+    /// the deadline is not met
+    pub fn arm<T>(&mut self, error: T)
+    where
+        T: std::future::Future + Send + 'static,
+        T::Output: Send + 'static,
+    {
+        let deadline = self.deadline;
+        let (s, mut r) = tokio::sync::mpsc::channel(1);
+        self.pet_chan = s;
+        tokio::spawn(async move {
+            let result = tokio::time::timeout(deadline, r.recv()).await;
+            if result.is_err() {
+                error.await;
+            }
+        });
+    }
+
+    /// meet the deadline
+    #[allow(dead_code)]
+    pub async fn pet(
+        &mut self,
+    ) -> Result<(), tokio::sync::mpsc::error::SendError<()>> {
+        self.pet_chan.send(()).await
+    }
+}
+
+#[derive(Default, Clone)]
+struct NodeStore {
+    inner: std::sync::Arc<NodeStoreInner>,
+}
+struct NodeStoreInner {
+    state: Mutex<HashMap<String, (Node, Watchdog)>>,
+    deadline: std::time::Duration,
+}
+impl Default for NodeStoreInner {
+    fn default() -> Self {
+        Self {
+            deadline: CliArgs::from_args().deadline.into(),
+            state: Default::default(),
+        }
+    }
+}
+
+impl NodeStore {
+    pub async fn register(&self, registration: Register) {
+        let mut state = self.inner.state.lock().await;
+
+        let mut watchdog = Watchdog::new(self.inner.deadline);
+        let id = registration.id.clone();
+        let store = self.clone();
+        let deadline = self.inner.deadline;
+        watchdog.arm(async move {
```

Comment:
just a small thing to consider is that as the number of nodes scales having a single timer that validates the deadline of all nodes might be useful?

Reply:
yeah I did start on something that stored the last time a node was seen online (gave up for laziness) so we could make use of that on a single watchdog which checks for timeouts for all nodes based on the "last seen" time.